design-docs/binder-tracing.md

```diff
@@ -0,0 +1,187 @@
+It seems that we should include begin/end traces for these operations regardless. An example of "folded" traces:
+```
+Begin AssemblyLoadContext.LoadFromAssemblyPath - assembly bind 12044
+    AssemblyLoadContext: App.CustomALC "My app ALC"
+    assembly path: /path/TestAssembly.dll
+```
+```
+End AssemblyLoadContext.LoadFromAssemblyPath - assembly bind 12044
+    AssemblyLoadContext: App.CustomALC "My app ALC"
+    Loaded assembly: "TestAssembly, version=1.2.3.0" location /path/TestAssembly.dll
+```
+
+## Sample diagnostic scenarios
+This section tries to describe common scenarios where binder diagnostics would help solve issues. It also tries to list the various pieces of information which the tracing should include to be helpful in the respective scenario.
+
+### Wrong version of assembly loaded
+If the assembly binding was successful but the wrong file was loaded there's no apparent place to store information about the binding results (like exceptions are for the failure cases). The tracing in the simple case would provide
+* What exact assembly was loaded (path, version, ...)
+* Who asked for the load (for example the assembly which had the assembly reference)
+* Where in the binder process we decided to look at the file which was loaded
+
+### Unexpected extension point inclusion
+The application implements some of the binder extension points, but the user is not aware of this, or is not expecting these extension points to be involved in the specific bind event. The tracing should include
+* Which extension points were called
+* Which one actually resolved the bind event
+* At least this information for all extension points regardless if they are aware of the tracing or not
+* Potentially more details from extension points which are aware of and use the tracing
+
+### Development of binder extension point
+The user is trying to implement an extension point for the binder (event handler, or custom ALC). Debugging the code without additional information is relatively hard since the code flows between the user code, low level framework and the runtime a lot. Call stacks can be confusing and order of execution is complicated. The tracing should help by providing:
+* The overall description of the bind event (what is trying to bind to what and so on)
+* Why the extension point was called and what has been tried before
+* What is the result of the extension point as seen by the binder
+* False positives - either unexpected successful bind or bind to the wrong thing outside of the extension point
+
+## Tracing mechanisms
+This section describes the proposed mechanisms to be used.
+### Context
+In order to provide the correlation of various traces which belong to the same bind event, there has to be a context which is accessible to all components which participate in the tracing. Maybe not at first but eventually we would like the tracing APIs to be available to 3rd party binding extensions (ALCs, event handlers).  
```

Comment:
The tracing API available to 3rd party code is EventSource. The 3rd party code can define their own events any way they like using EventSource. I do not think we need to have a separate public tracing API for binding.

Reply:
The reason for the addition would be the ability to get access to the "bind ID" so that 3rd party events would correlate with 1st party events the same way.
Potentially also to get access to additional information about the current bind operation for tracing purposes (although this might be tricky as people could start using this for functional differences as well).
And also if it would make sense to have everybody to use the same event type for a given thing... not sure if there's such an event type yet.

That said - in the first iteration I definitely think we would not expose the API just yet.

Reply:
Thanks for pointing out EventSource activity IDs to me. I rewrote the mechanics section to align with activity IDs and their handling.